File: baekjoon/dfs&bfs/11724.py
import sys
from collections import deque

input = sys.stdin.readline

N, M = map(int,input().split())

graph = [[] for _ in range(N+1)]
isVisited = [False] * (N+1)

for _ in range(M):
    startNode, endNode = map(int,input().split())
    graph[startNode].append(endNode)
    graph[endNode].append(startNode)

def dfs(graph, startNode, isVisited):

    isVisited[startNode] = True
    print(startNode)
    for i in graph[startNode]:
        if isVisited[i] == False:
            dfs(graph,i,isVisited)

# bfs만 되네...??
def bfs(graph, startNode, isVisited):
    q = deque()
    q.append(startNode)
    isVisited[startNode] = True

    while q:
        node = q.popleft()
        print(node)
        for i in graph[node]:
            if isVisited[i] == False:
                isVisited[i] = True
                q.append(i)

cost = 0
for i in range(1,N+1):

    if isVisited[i] == False:
        cost += 1
        bfs(graph,i,isVisited)

print("cost: %d" %cost)

File: baekjoon_js/dfs&bfs/11724.js
("use strict");

const fs = require("fs");
const stdin = (process.platform === "linux"
  ? fs.readFileSync("/dev/stdin").toString()
  : `6 8
1 2
2 5
5 1
3 4
4 6
5 4
2 4
2 3
`
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [N, M] = input().split(" ").map(Number);
let graph = Array(N + 1)
  .fill(null)
  .map(() => Array());

let isVisited = Array(N + 1).fill(false);

for (let i = 0; i < M; i++) {
  const [startNum, endNum] = input().split(" ").map(Number);
  graph[startNum].push(endNum);
  graph[endNum].push(startNum);
}

cnt = 0;
for (let i = 1; i < N; i++) {
  if (!isVisited[i]) {
    dfs(i);
    cnt += 1;
  }
}

console.log(cnt);

function dfs(start) {
  isVisited[start] = true;
  for (const nextNode of graph[start]) {
    if (!isVisited[nextNode]) {
      dfs(start);
    }
  }
}

function bfs(start) {
  q = [];
  q.push(start);

  while (q.length > 0) {
    node = q.shift();
    isVisited[node] = true;

    for (const nextNode of graph[node]) {
      //   console.log(nextNode);
      if (!isVisited[nextNode]) {
        q.push(nextNode);
      }
    }
  }
}
